pkg/configmap/configmap_test.go

```diff
@@ -1 +1,322 @@
+		nsname      string
+		expectedCM  *corev1.ConfigMap
+		expectedErr string
+	}{
+		{
+			name:   "test",
+			nsname: "testns",
+			expectedCM: &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test",
+					Namespace: "testns",
+				},
+			},
+			expectedErr: "",
+		},
+		{
+			name:        "",
+			nsname:      "testns",
+			expectedCM:  nil,
+			expectedErr: "configmap 'name' cannot be empty",
+		},
+		{
+			name:        "test",
+			nsname:      "",
+			expectedCM:  nil,
+			expectedErr: "configmap 'nsname' cannot be empty",
+		},
+	}
+
+	for _, testCase := range testCases {
+		testSettings := clients.GetTestClients([]runtime.Object{})
+
+		testBuilder := NewBuilder(testSettings, testCase.name, testCase.nsname)
+
+		if testCase.expectedErr == "" {
+			assert.NotNil(t, testBuilder)
+			assert.Equal(t, testCase.expectedCM.Name, testBuilder.Definition.Name)
+			assert.Equal(t, testCase.expectedCM.Namespace, testBuilder.Definition.Namespace)
+		} else {
+			assert.Equal(t, testCase.expectedErr, testBuilder.errorMsg)
+		}
+	}
+}
+
+func TestPull(t *testing.T) {
+	testCases := []struct {
+		name                string
+		nsname              string
+		expectedError       bool
+		addToRuntimeObjects bool
+		expectedErrorText   string
+	}{
+		{
+			name:                "test",
+			nsname:              "testns",
+			expectedError:       false,
+			addToRuntimeObjects: true,
+		},
+		{
+			name:                "",
+			nsname:              "testns",
+			expectedError:       true,
+			addToRuntimeObjects: false,
+			expectedErrorText:   "configmap 'name' cannot be empty",
+		},
+		{
+			name:                "test",
+			nsname:              "",
+			expectedError:       true,
+			addToRuntimeObjects: false,
+			expectedErrorText:   "configmap 'nsname' cannot be empty",
+		},
+		{
+			name:                "test",
+			nsname:              "testns",
+			expectedError:       true,
+			addToRuntimeObjects: false,
+			expectedErrorText:   "configmap object test doesn't exist in namespace testns",
+		},
+	}
+
+	for _, testCase := range testCases {
+		var (
+			runtimeObjects []runtime.Object
+			testSettings   *clients.Settings
+		)
+
+		testCM := generateConfigMap(testCase.name, testCase.nsname)
+
+		if testCase.addToRuntimeObjects {
+			runtimeObjects = append(runtimeObjects, testCM)
+		}
+
+		testSettings = clients.GetTestClients(runtimeObjects)
+
+		// Test the Pull function
+		builderResult, err := Pull(testSettings, testCase.name, testCase.nsname)
+
+		// Check the error
+		if testCase.expectedError {
+			assert.NotNil(t, err)
+
+			if testCase.expectedErrorText != "" {
+				assert.Equal(t, testCase.expectedErrorText, err.Error())
+			}
+		} else {
+			assert.Nil(t, err)
+			assert.NotNil(t, builderResult)
+			assert.Equal(t, testCase.name, builderResult.Definition.Name)
+			assert.Equal(t, testCase.nsname, builderResult.Definition.Namespace)
+		}
+	}
+}
+
+func buildTestBuilderWithFakeObjects(objects []runtime.Object) *Builder {
```

Comment:
Please move private function to the end of this file.

Reply:
Done.